aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,266 @@
+                                        Some(pending_keypair) => {
+                                            // Pending key is present. Are we ready to promote it
+                                            // to active?
+                                            if clock
+                                                .now()
+                                                .difference(pending_keypair.updated_at())
+                                                .unwrap()
+                                                > hpke_key_rotator_config.pending_duration
+                                            {
+                                                tx.set_global_hpke_keypair_state(
+                                                    pending_keypair.hpke_keypair().config().id(),
+                                                    &HpkeKeyState::Active,
+                                                )
+                                                .await?;
+                                                tx.set_global_hpke_keypair_state(
+                                                    active_keypair.hpke_keypair().config().id(),
+                                                    &HpkeKeyState::Expired,
+                                                )
+                                                .await?;
+                                            }
+                                        }
+                                        None => {
+                                            // No pending key is present but the active key is aged
+                                            // out. Insert one.
+                                            tx.put_global_hpke_keypair(
+                                                &generate_hpke_config_and_private_key(
+                                                    next_available_id()?,
+                                                    ciphersuite.kem_id(),
+                                                    ciphersuite.kdf_id(),
+                                                    ciphersuite.aead_id(),
+                                                )
+                                                .map_err(|e| DatastoreError::User(e.into()))?,
+                                            )
+                                            .await?;
+                                        }
+                                    }
+                                }
+                            }
+                            None => {
+                                // Bootstrapping case: there is no active keypair with the given
+                                // ciphersuite. This means one of the following cases:
+                                //   - Fresh instance of Janus coming up, and a new key needs to be
+                                //     bootstrapped.
+                                //   - The operator has added a new ciphersuite to the config.
```

Comment:
Technically in this case, we would want to go through the usual delay with the key in a pending state. That would either take more aggressive checks in the upload path, which may not be wise, or Tink-style feedback of which keys are in use where, which we don't have appropriate infrastructure for in our deployment model. We could just document that you have to create a pending global keypair with a given ciphersuite, and wait, before you can add that ciphersuite to the key rotator's configuration.

Reply:
Maybe we could make the step where we immediately set the state to active conditional on `current_keypairs` being empty. Then, it would only come into play in the bootstrapping scenario or in the case of outside interference.